flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/reader/PulsarSourceReaderFactory.java

```diff
@@ -0,0 +1,160 @@
+                                sourceConfiguration.getCloseTimeoutMs()));
+        splitCloser.register(
+                () ->
+                        closeWithTimeout(
+                                "PulsarAdmin",
+                                pulsarAdmin::close,
+                                sourceConfiguration.getCloseTimeoutMs()));
+
+        // Create different pulsar source reader by subscription type.
+        SubscriptionType subscriptionType = sourceConfiguration.getSubscriptionType();
+        if (subscriptionType == SubscriptionType.Failover
+                || subscriptionType == SubscriptionType.Exclusive) {
+            // Create a ordered split reader supplier.
+            Supplier<PulsarOrderedPartitionSplitReader<IN, OUT>> splitReaderSupplier =
+                    () -> {
+                        PulsarOrderedPartitionSplitReader<IN, OUT> reader =
+                                new PulsarOrderedPartitionSplitReader<>(
+                                        pulsarClient,
+                                        pulsarAdmin,
+                                        configuration,
+                                        sourceConfiguration,
+                                        consumerConfigurationCustomizer,
+                                        deserializationSchema);
+
+                        splitCloser.register(reader);
+                        return reader;
+                    };
+
+            return new PulsarOrderedSourceReader<>(
+                    elementsQueue,
+                    splitReaderSupplier,
+                    configuration,
+                    readerContext,
+                    sourceConfiguration,
+                    splitCloser);
+        } else if (subscriptionType == SubscriptionType.Shared
+                || subscriptionType == SubscriptionType.Key_Shared) {
+            TransactionCoordinatorClient coordinatorClient =
+                    ((PulsarClientImpl) pulsarClient).getTcClient();
+            if (coordinatorClient == null && !sourceConfiguration.isEnableAutoCommitCursor()) {
+                throw new PulsarRuntimeException("Transaction is required but didn't enabled");
+            }
+
+            Supplier<PulsarUnorderedPartitionSplitReader<IN, OUT>> splitReaderSupplier =
+                    () -> {
+                        PulsarUnorderedPartitionSplitReader<IN, OUT> reader =
+                                new PulsarUnorderedPartitionSplitReader<>(
+                                        pulsarClient,
+                                        pulsarAdmin,
+                                        configuration,
+                                        sourceConfiguration,
+                                        consumerConfigurationCustomizer,
+                                        deserializationSchema,
+                                        coordinatorClient);
+
+                        splitCloser.register(reader);
```

Comment:
The `PulsarUnorderedPartitionSplitReader` should  already be closed by the `SplitFetcherManager`.